hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemChooseSAS.java

```diff
@@ -0,0 +1,145 @@
+  public void setup() throws Exception {
+    createFilesystemForSASTests();
+    super.setup();
+    // obtaining an account SAS token from in-built generator to set as configuration for testing filesystem level operations
+    generateAccountSAS();
+  }
+
+  /**
+   * Tests the scenario where both the token provider class and a fixed token are configured:
+   * whether the correct choice is made (precedence given to token provider class), and the chosen SAS Token works as expected
+   * @throws Exception
+   */
+  @Test
+  public void testBothProviderFixedTokenConfigured() throws Exception {
+    AbfsConfiguration testAbfsConfig = getConfiguration();
+
+    // configuring a SASTokenProvider class: this provides a user delegation SAS
+    // user delegation SAS Provider is set
+    // This easily distinguishes between results of filesystem level and blob level operations to ensure correct SAS is chosen,
+    // when both a provider class and fixed token is configured.
+    testAbfsConfig.set(FS_AZURE_SAS_TOKEN_PROVIDER_TYPE, "org.apache.hadoop.fs.azurebfs.extensions.MockDelegationSASTokenProvider");
+
+    // configuring the fixed SAS token
+    testAbfsConfig.set(FS_AZURE_SAS_FIXED_TOKEN, accountSAS);
+
+    // creating a new fs instance with the updated configs
+    AzureBlobFileSystem newTestFs = (AzureBlobFileSystem) FileSystem.newInstance(testAbfsConfig.getRawConfiguration());
+
+    // testing a file system level operation
+    TracingContext tracingContext = getTestTracingContext(newTestFs, true);
+    // expected to fail in the ideal case, as delegation SAS will be chosen, provider class is given preference when both are configured
+    // this expectation is because filesystem level operations are beyond the scope of Delegation SAS Token
+    intercept(SASTokenProviderException.class,
+        () -> {
+          newTestFs.getAbfsStore().getFilesystemProperties(tracingContext);
+        });
+
+    // testing blob level operation to ensure delegation SAS token is otherwise valid and above operation fails only because it is fs level
+    Path testPath = new Path("/testCorrectSASToken");
+    newTestFs.create(testPath).close();
+  }
+
+  /**
+   * Tests the scenario where only the fixed token is configured, and no token provider class is set:
+   * whether fixed token is read correctly from configs, and whether the chosen SAS Token works as expected
+   * @throws IOException
+   */
+  @Test
+  public void testOnlyFixedTokenConfigured() throws IOException {
+    AbfsConfiguration testAbfsConfig = getConfiguration();
+
+    // clearing any previously configured SAS Token Provider class
+    testAbfsConfig.unset(FS_AZURE_SAS_TOKEN_PROVIDER_TYPE);
+
+    // setting an account SAS token in the fixed token field
+    testAbfsConfig.set(FS_AZURE_SAS_FIXED_TOKEN, accountSAS);
+
+    // creating a new FS with updated configs
+    AzureBlobFileSystem newTestFs = (AzureBlobFileSystem) FileSystem.newInstance(testAbfsConfig.getRawConfiguration());
```

Comment:
this needs to be closed() after use, so use in a try-with-resources clause

Reply:
Taken